docs/decisions/api-browser-security.md

```diff
@@ -0,0 +1,80 @@
+
+The backend will need to periodically refresh the session cookie and check that the session has not been revoked by the user in Keycloak (e.g. using the [GET /admin/realms/{realm}/users/{user-id}/sessions](https://www.keycloak.org/docs-api/latest/rest-api/index.html#_get_adminrealmsrealmusersuser_idsessions) API)
+
+User Agent Filtering could be implemented to ensure that mobile apps don't inadvertently use a WebView in the authentication flow.
+
+### Confirmation
+
+Re-user clients should be created via a standard tool so that permissions are set correctly, depending on the type of client.
+
+## Pros and Cons of the Options
+
+### Allow Access Tokens to be used in all API requests
+
+With this approach we make it the re-user's responsibility to secure their Access Tokens. Our APIs would support CORS so that API calls can be made directly from the browser. The consequence of this is that the re-user may choose to store the Access Token in an insecure way.
+
+This can be mitigated to some extent by enabling the [Revoke Refresh Token](https://www.keycloak.org/docs/latest/server_admin/index.html#_offline-access) option so that a Refresh Token can only be used once. However, there are still attacks that can get around this.
+
+* Good: Simplest to implement. API calls are the same whether from a web application, mobile application or backend
+* Neutral: The client would only be able to call authenticated APIs on behalf of a user, so user consent would always be required
+* Bad: Open to XSS attacks or others depending on how the Access / Refresh Token is stored
+
+### Only allow Authenticated APIs to be called from backends
+
+This could be achieved by disabling CORS on our authenticated APIs and only creating non-public clients in Keycloak, thus preventing Implicit or PKCE login flows.
+
+* Good: There is no incentive for a re-user to store an Access Token in the browser
+* Neutral: Increased latency (but this only applies to authenticated actions like product updates where the frequency is low)
+* Bad: Web applications will always need a corresponding backend to obtain Access Tokens and call authenticated APIs
+
+### Implement CORS-enabled authentication endpoint to start a session
+
+With this approach, a web application would go through a normal PKCE flow to obtain an access token and then use this to call a CORS-enabled endpoint (analogous to the existing Login API) to create a session, which would be stored in a Secure HttpOnly cookie. The web application can then forget the access token as there is no need to use this on subsequent requests. Other APIs would disable CORS so could only be called with a session cookie from a browser, but could still be called with an access token from a mobile app or backend.
+
+* Good: Access tokens are not retained in the browser
+* Neutral: An additional step is required to start a session
```

Comment:
I'd classify this as _potentially_ bad, actually.

1. With this general proposal, we likely also want the Robotoff API to be secured the same way, which would mean that Robotoff would either have to implement the same session management as ProductOpener, or would have to use PO's session management. I might be mistaken, but that would contradict some of the advantages that adding Keycloak has the first place?
2. Assuming that ProductOpener wouldn't accept the access token for APIs other than the session management, this removes the possibility of using well-tested OOTB OIDC/OAuth2 libraries.
3. Additional overhead of implementing and maintaining session management. It's often said that rolling your own security should be done carefully.

Reply:
The secure cookie can be tied to the base domain, e.g .openfoodfacts.org so could work with other subdomains. We could even add the login API to the Keycloak code potentially so it isn't tied to any other service